packages/slider/src/useSlider.ts

```diff
@@ -0,0 +1,364 @@
+/**
+ * Copyright Zendesk, Inc.
+ *
+ * Use of this source code is governed under the Apache License, Version 2.0
+ * found at http://www.apache.org/licenses/LICENSE-2.0.
+ */
+
+import { useCallback, useEffect, useMemo, useReducer, useRef, useState } from 'react';
+import { composeEventHandlers, KEYS } from '@zendeskgarden/container-utilities';
+import throttle from 'lodash.throttle';
+
+import { IUseSliderProps, IUseSliderReturnValue } from './types';
+import {
+  sliderReducer,
+  initializeSliderReducerState,
+  getThumbCurrentValueNumber,
+  getThumbMaxValueNumber,
+  getThumbMinValueNumber,
+  setThumbValue,
+  stepUp,
+  stepDown,
+  resetRangeMin,
+  resetRangeMax,
+  DEFAULT_MIN,
+  DEFAULT_MAX,
+  DEFAULT_STEP
+} from './sliderReducer';
+
+const POSSIBLE_SLIDER_KEYS = [KEYS.LEFT, KEYS.RIGHT, KEYS.UP, KEYS.DOWN, KEYS.HOME, KEYS.END];
+
+export function useSlider({
+  defaultValue,
+  min = DEFAULT_MIN,
+  max = DEFAULT_MAX,
+  step = DEFAULT_STEP,
+  rtl = false,
+  required,
+  disabled,
+  readOnly
+}: IUseSliderProps): IUseSliderReturnValue {
+  const trackElement = useRef<Element>(null);
+  const [trackElementDimensions, setTrackElementDimensions] = useState<DOMRect | null>(null);
+  const isInteractive = disabled === false && readOnly === false;
+  const [slidingThumbIndex, setSlidingThumbIndex] = useState<number | null>(null);
+  const [state, dispatch] = useReducer(
+    sliderReducer,
+    initializeSliderReducerState(defaultValue || [min, max])
+  );
+
+  /**
+   * Set up a ResizeObserver to capture the track element's dimensions,
+   * even when those dimensions, even when those dimensions change.
+   */
+  useEffect(() => {
+    if (trackElement.current === null) {
+      return;
+    }
+
+    // we need to check the scope/runtime and make sure it has resize observer
+    // we should also evaluate if a library may be a better solution than using a "newish" browser API
```

Comment:
For reference, here's the [Can I Use table](https://caniuse.com/resizeobserver) for the ResizeObserver API. It has pretty good support across desktop and mobile. 👍

Reply:
@jgorfine-zendesk is there any pre-existing library that would handle this for us and provide a polyfill or backfall if `ResizeObserver` were not present?